### PR TITLE
Added Mozjpeg Optimizer

### PR DESCRIPTION
**Pull request by cfreak**
Added Mozjpeg Optimizer for a better JPEG Optimization.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -57,6 +58,18 @@ sudo npm install -g svgo
 sudo apt-get install gifsicle
 ```
 
+Mozjpeg for Ubuntu:
+```bash
+sudo apt-get install autoconf automake libtool nasm make pkg-config git
```
Comment (freekmurze) [inline README.md]: That's a lot of work to install this. Is there a way to install it with one simple `apt-get`?

Comment (cfreak) [inline README.md]: No sorry, i didn't find a way to make it easier.

**Comment by freekmurze** [comment]
Seems like mozjpeg is a bit too unstable atm. 

https://twitter.com/freekmurze/status/889767525592035328

Closing this for now, will reconsider this in the future

File: README.md (hunk 1 above)
Comment (BlackIkeEagle) [inline README.md]: You could ship a statically built binary for the OS'es you support but that is an extra maintenance burden you might not want

**Comment by fgilio** [comment]
@freekmurze would you be open to revisit this? I've using mozjpeg in production for over 2 years, no problem at all. I actually just found out that Instagram also uses mozjpeg, and I image lots of other enterprises do too.

File: README.md (hunk 1 above)
Comment (tingan) [reply README.md:63]: for ubuntu and debian, this is a deb package available, https://mozjpeg.codelove.de/binaries.html.

**Comment by freekmurze** [comment]
@fgilio I could reconsider this. Is there already a better we to install it on Ubuntu on Mac?

**Comment by mihaikelemen** [comment]
It still requires compliation from the source. Maybe adding a small guide on how to accomplish (building from source) this will help people adopting mozjpeg.

**Comment by fgilio** [comment]
I spent a lot of time searching for this last week and just couldn't find it.

We ended up implementing this:
```php
use Spatie\ImageOptimizer\Image;
use Spatie\ImageOptimizer\Optimizers\BaseOptimizer;

class SpatieImageOptimizerMozjpeg extends BaseOptimizer
{
    /*
     * mozjpeg comes with 2 binaries that work with images:
     * - cjpeg does lossy compression
     * - jpegtran does lossless compression
     *
     * Refs:
     * - https://stackoverflow.com/a/48123459
     * - https://github.com/mozilla/mozjpeg/issues/229
     */
    public $binaryName = 'cjpeg';

    public function canHandle(Image $image): bool
    {
        return $image->mime() === 'image/jpeg';
    }

    public function getCommand(): string
    {
        $binaryPath = config('settings.mozjpeg_bins_dir_path');
        $optionString = implode(' ', $this->options);

        $escapedImagePath = escapeshellarg($this->imagePath);
        $escapedTmpImagePath = escapeshellarg($this->imagePath.'tmp');
        /*
         * mozjpeg won't overwrite the source image, so we create a temporary one as the output for the optimization
         * and then manually overwrite the original with the temporary
         */
        $command = config('settings.mozjpeg_bins_dir_path').$this->binaryName." {$optionString} {$escapedImagePath} > {$escapedTmpImagePath} && mv -f {$escapedTmpImagePath} {$escapedImagePath}";

        logger()->debug('SpatieImageOptimizerMozjpeg', [
            'command_line' => $command,
        ]);

        return $command;
    }
}
```

And then this to get mozjpeg working in Alpine Linux:
```dockerfile
## To get mozjpeg working
# Source: https://jamesloper.com/how-to-build-mozjpeg
# We removed deps already installed or that are availabe by default with Vapor's image https://github.com/laravel/vapor-dockerfiles/blob/master/php74.Dockerfile
RUN apk add --no-cache cmake automake libtool nasm make pkgconfig
RUN git clone https://github.com/mozilla/mozjpeg.git
# We were having problens with PNG compatibility, `-DENABLE_STATIC=FALSE -DPNG_SUPPORTED=TRUE` fixes it
# Source: https://github.com/mozilla/mozjpeg/issues/351#issuecomment-827121956
RUN cd mozjpeg && mkdir build && cd build && sudo cmake -G"Unix Makefiles" -DENABLE_STATIC=FALSE -DPNG_SUPPORTED=TRUE ../ && sudo make install
```

Hope it helps someone!

**Pull request closed without merging**